### PR TITLE
TypeScript Fetch:  implementation

### PR DESCRIPTION
**Pull request by leonyu**
TypeScript implementation using [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)

Other features: 
* uses ES6 modules system (backward compatible to CommonJS)
* provides barebone `package.json` so user can `npm install` the dependencies
* provides `tsconfig.json` so user can `tsc`
* provides `typings.json` so user can install TypeScript definition for dependencies via `typings` (the prevailing type management system. 
* code is isomorphic for both Node and browser (to be fully tested) 

Limitations: 
* it does not support authentication currently

### Checklist ###
* [ ] Update `package.json` so the compiled module would be readily usable as `npm install {codegen-client-path/git}` or `npm link {codegen-client-path}` or other user-friendly methods
* [ ] Sample usage with `petstore` app
* [ ] Manual test
* [ ] Integration test?

### Test Cases ###
* Environment/bundling (isomorphic)
 * [ ] Node.js
 * [ ] Webpack
 * [ ] Browserify (TBD)
* Module system compatibility
 * [ ] TypeScript ES6 module syntax
 * [ ] TypeScript CommonJS/AMD module syntax
 * [ ] JavaScript CommonJS module syntax

### CODE REVIEW

**Comment by wing328** [comment]
@leonyu thanks for the PR and detailed explanation provided above. The change looks good to me.

cc @Vrolijkx to see if he's any feedback.

File: modules/swagger-codegen/src/main/resources/TypeScript-Fetch/package.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (Vrolijkx) [inline modules/swagger-codegen/src/main/resources/TypeScript-Fetch/package.json]: Don't we need to have an install script to do tsc and install typings? see https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/typescript-angular2/package.mustache for an example.

**Comment by Vrolijkx** [comment]
@leonyu Nice job creating a client that can both be used on node and browser .

File: modules/swagger-codegen/src/main/resources/TypeScript-Fetch/package.json (hunk 1 above)
Comment (leonyu) [inline modules/swagger-codegen/src/main/resources/TypeScript-Fetch/package.json]: Fixed for now

**Comment by leonyu** [comment]
I made a checklist of what I expect to do before having this Ready for Merge. Mostly on usability and testing. 

My goal for this template is that I want it to build to a private NPM module that is readily usable in Node by TypeScript and JavaScript. It should be isomorphic in it can be bundled by Webpack (and Browserify as well if the Node dependencies are compatible). 

In addition, what is the normal process for getting new templates merged? Do you guys have your own checklist, tests, validations that I should be aware of?

**Comment by wing328** [comment]
@leonyu the checklist looks good.

For your goal, please let us (the community) know if there's any help you need from us to achieve that :)

About the normal process, currently, we only have this wiki page: https://github.com/swagger-api/swagger-codegen/wiki/How-to-add-a-generator-for-a-new-language-or-framework

We'll incorporate part of your check list to the page so that new contributors will have a better idea what we expect for a new client/sever generator.

**Comment by wing328** [comment]
FYI. Checklist added: https://github.com/swagger-api/swagger-codegen/wiki/How-to-add-a-generator-for-a-new-language-or-framework#checklist-a-list-of-items-we-look-for-in-the-new-generator

**Comment by wing328** [comment]
@leonyu I've merged the changes into master with some minor enhancements:
- shell script and batch file to generate TypeScript-Fetch Petstore sample
- Add the word "Beta" to help message as the typescript-fetch client lacks authentication support at the moment

Thank you again for your PR. I'm sure TypeScript developers appreciate your contribution.

**Comment by leonyu** [comment]
Cool thank you very much for the merge, I was actually working on the unit tests, I will make a follow up PR for that.

**Pull request merged**